### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: src/contract.ts [hunk 1]
```diff
@@ -165,6 +318,19 @@ export class AbstractContract {
         // in vscode termianal need to use [:] to jump to file line, but here need to use [#] to jump to file line in output channel.
         if (opcode.pos) {
           error = `VerifyError: ${bsi.errstr} \n\t[Go to Source](${path2uri(opcode.pos.file)}#${opcode.pos.line})  fails at ${opcode.opcode}\n`;
+
+          if (args && ['OP_CHECKSIG', 'OP_CHECKSIGVERIFY'].includes(opcode.opcode)) {
+            if (!txCtx) {
+              throw new Error('should provoid txContext when verify');
```
Comment (xhliu) [inline src/contract.ts:324]: provoid -> provide

File: src/contract.ts [hunk 2]
```diff
@@ -165,6 +318,19 @@ export class AbstractContract {
...
+            } if (!tx) {
+              throw new Error('should provoid txContext.tx when verify');
```
Comment (xhliu) [inline src/contract.ts:326]: same

File: src/contract.ts [hunk 3]
```diff
@@ -165,6 +318,19 @@ export class AbstractContract {
         // in vscode termianal need to use [:] to jump to file line, but here need to use [#] to jump to file line in output channel.
         if (opcode.pos) {
           error = `VerifyError: ${bsi.errstr} \n\t[Go to Source](${path2uri(opcode.pos.file)}#${opcode.pos.line})  fails at ${opcode.opcode}\n`;
+
+          if (args && ['OP_CHECKSIG', 'OP_CHECKSIGVERIFY'].includes(opcode.opcode)) {
```
Comment (xhliu) [inline src/contract.ts:322]: Also for multisig
'OP_CHECKMULTISIG', 'OP_CHECKMULTISIGVERIFY'

File: src/contract.ts [hunk 4]
```diff
@@ -80,26 +81,178 @@ export class AbstractContract {
...
+          this.lockingScript.toASM(),
+          txCtx.inputSatoshis,
+          txCtx.inputIndex,
+          parseInt(sig.slice(sig.length - 2, sig.length), 16),
+          DEFAULT_FLAGS
+        );
+
+        const title = '----- CheckSig Fail Hints Begin -----';
+        const body = [
+          'You should make sure the following check points all passed:',
```
Comment (xhliu) [inline src/contract.ts:145]: 'You should make sure the following checkpoints all pass:'

File: test/gotosource.test.ts [hunk 5]
```diff
@@ -161,12 +161,27 @@ describe('VerifyError', () => {
     });
 
 
+    it('stop tokenUtxo.scrypt#14 => require(Tx.checkPreimage(txPreimage))', () => {
```
Comment (xhliu) [inline test/gotosource.test.ts:164]: Also add a test for checkSig() failure

**Review by xhliu — APPROVED** [review]
:shipit: great!

File: test/gotosource.test.ts (hunk 5 above)
Comment (zhfnjust) [reply test/gotosource.test.ts:164]: already have  at 
unit test stop at p2pkh.scrypt#10

**Pull request merged**